src/gf.c

```diff
@@ -1552,7 +1552,7 @@ jl_method_instance_t *jl_method_lookup_by_type(jl_methtable_t *mt, jl_tupletype_
         JL_UNLOCK(&mt->writelock);
         return linfo;
     }
-    if (jl_is_leaf_type((jl_value_t*)types))
+    if (jl_is_leaf_type((jl_value_t*)types)) // FIXME: this is the wrong predicate
```

Comment:
Concerning... could you elaborate on why it's the wrong predicate, why you can't use the right predicate and in what direction this would need to change in order to be correct?

Reply:
#23606